src/System.Text.Json/src/System/Text/Json/Serialization/JsonClassInfo.cs

```diff
@@ -66,85 +77,104 @@ internal JsonClassInfo(Type type, JsonSerializerOptions options)
         {
             ulong key = GetKey(propertyName);
             JsonPropertyInfo info = null;
 
             // First try sorted lookup.
-            int count = _property_refs_sorted.Count;
-            if (count != 0)
+            int propertyIndex = frame.PropertyIndex;
+
+            // If we're not trying to build the cache locally and there is an existing cache then use it.
+            bool hasPropertyCache = frame.PropertyRefCache == null && _propertyRefsSorted != null;
+            if (hasPropertyCache)
             {
-                int iForward = propertyIndex;
-                int iBackward = propertyIndex - 1;
-                while (iForward < count || (iBackward >= 0 && iBackward < count))
+                int count = _propertyRefsSorted.Length;
+                if (count != 0)
                 {
-                    if (iForward < count)
+                    int iForward = propertyIndex;
+                    int iBackward = propertyIndex - 1;
+                    while (iForward < count || (iBackward >= 0 && iBackward < count))
                     {
-                        if (TryIsPropertyRefEqual(_property_refs_sorted, propertyName, key, iForward, out info))
+                        if (iForward < count)
                         {
-                            return info;
+                            if (TryIsPropertyRefEqual(ref _propertyRefsSorted[iForward], propertyName, key, out info))
+                            {
+                                return info;
+                            }
+                            ++iForward;
                         }
-                        ++iForward;
-                    }
 
-                    if (iBackward >= 0)
-                    {
-                        if (TryIsPropertyRefEqual(_property_refs_sorted, propertyName, key, iBackward, out info))
+                        if (iBackward >= 0)
                         {
-                            return info;
+                            if (TryIsPropertyRefEqual(ref _propertyRefsSorted[iBackward], propertyName, key, out info))
+                            {
+                                return info;
+                            }
+                            --iBackward;
                         }
-                        --iBackward;
                     }
                 }
             }
 
-            // Then try fallback
-            for (int i = 0; i < _property_refs.Count; i++)
+            // Try the main list which has all of the properties.
+            for (int i = 0; i < _propertyRefs.Count; i++)
             {
-                if (TryIsPropertyRefEqual(_property_refs, propertyName, key, i, out info))
+                PropertyRef propertyRef = _propertyRefs[i];
+                if (TryIsPropertyRefEqual(ref propertyRef, propertyName, key, out info))
                 {
                     break;
                 }
             }
 
-            if (info != null)
+            if (!hasPropertyCache)
```

Comment:
Can this be an else condition to `if (hasPropertyCache)` on line 90, before the loop?

Reply:
No. I added a comment to explain why (race conditions).